src/include/migraphx/match/gelu_tanh.hpp

```diff
@@ -13,23 +13,73 @@ template <class F>
 struct gelu_tanh_matcher
 {
     F f;
-    auto pow_fn() const { return f("pow")(used_once(), arg(1)(has_value(3.0f))); }
 
-    auto tanh_fn() const
+    // helper function matches "pow" operator and 3.0
+    auto pow_fn() const { return f("pow")(used_once(), arg(0).bind("x"), arg(1)(has_value(3.0f))); }
+
+    // clang-format off
+    // helper function matches the subexpression of the gelu formula:
+    // tanh(sqrt(2 / pi) * (x + 0.044715 * pow(x, 3)))
+    // clang-format on
+    // This is the expression that is actually simplified to the erf function.
+    // Also binds the name "x" to the first appearance of the x argument.
+    // Note that there is no checking that the "x" here matches the two other appearances
+    // of x in the gelu formula.
+    auto tanh_sub_fn() const
     {
-        return f("tanh")(
-            used_once(),
-            arg(0)(f("mul")(either_arg(0, 1)(has_value(sqrt(M_2_PI), 1e-3),
-                                             f("add")(any_arg(0, 1)(f("mul")(either_arg(0, 1)(
-                                                 has_value(0.044715f), pow_fn()))))))));
+        // magic number * x**3
+        auto const_times_pow_fn = f("mul")(either_arg(0, 1)(has_value(0.044715f), pow_fn()));
+        // add x to result of const_times_pow_fn
+        // The instruction matched by any() here should be the same x that was bound in pow_fn(),
+        // but matcher has no way to ensure they're the same.  Do not bind to label "x" a second
+        // time as it leads to undefined results.
+        auto add_x_fn = f("add")(either_arg(0, 1)(any(), const_times_pow_fn));
+
+        // multiply by sqrt(2 / pi)
+        auto mul_sqrt_2_over_pi =
+            f("mul")(either_arg(0, 1)(add_x_fn, has_value(sqrt(M_2_PI), 1e-3)));
+
+        // tanh
+        auto tanh_fn = f("tanh")(used_once(), arg(0)(mul_sqrt_2_over_pi));
+        return tanh_fn;
     }
 
     auto matcher() const
     {
-        return f("mul")(used_once(),
-                        either_arg(0, 1)(any().bind("x"),
-                                         f("add")(any_arg(0, 1)(f("mul")(
-                                             either_arg(0, 1)(has_value(0.5f), tanh_fn()))))));
+        // clang-format off
+        // Formula used in gelu.cpp:   0.5 * x * (1 + tanh(sqrt(2 / pi) * (x + 0.044715 * pow(x, 3))))
+        //      but we only match the subexpression without the 0.5 factor:
+        //      x * (1 + tanh(sqrt(2 / pi) * (x + 0.044715 * pow(x, 3))))
+        //      because that's how it appears in common bert models.
+        //
+        // matcher grammar:
```

Comment:
I would prefer these grammar comments go in matcher.hpp. Or if there are portions that are specific to this matcher, you can point them out. Otherwise, a simple comment to refer to matcher.hpp would suffice.